### PR TITLE
[tune] New Doc edits, add Concepts page

### PR DESCRIPTION
**Pull request by richardliaw**
## Why are these changes needed?

Added concepts page, moved examples and usage guide into one "Learning"
folder. 

TODO next:

1. Remove Trial Scheduler and Search Algorithm pages.
1a. Add a matrix for what things are compatible.
2. Move Logging/Debugging/Troubleshooting out?
2a. Add "how to use wandb" or log within trainable.
3. Add a FAQ
4. Add a guide for how to optimize hyperparameters/use a trial
scheduler.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

**Comment by anabranch** [comment]
Is anything autogenerated here @richardliaw ?

**Comment by richardliaw** [comment]
Nope, not in the diff.

File: doc/source/tune/api_docs/trainable.rst [hunk 1]
```diff
@@ -109,6 +109,28 @@ Use ``validate_save_restore`` to catch ``_save``/``_restore`` errors before exec
     validate_save_restore(MyTrainableClass)
     validate_save_restore(MyTrainableClass, use_object_store=True)
 
+
+Advanced Resource Allocation
+----------------------------
+
+Trainables can themselves be distributed. If your trainable function / class creates further Ray actors or tasks that also consume CPU / GPU resources, in ``tune.run``, you will also want to set ``extra_cpu`` or ``extra_gpu`` to reserve extra resource slots. For example, if a trainable class requires 1 GPU itself, but will launch 4 actors each using another GPU, then it should set ``"gpu": 1, "extra_gpu": 4``.
```
Comment (sven1977) [inline doc/source/tune/api_docs/trainable.rst:116]:
```suggestion
Trainables can themselves be distributed. If your trainable function / class creates further Ray actors or tasks that also consume CPU / GPU resources, you will want to set ``extra_cpu`` or ``extra_gpu`` inside ``tune.run`` to reserve extra resource slots. For example, if a trainable class requires 1 GPU itself, but also launches 4 actors, each using another GPU, then you should set ``"gpu": 1, "extra_gpu": 4``.
```

File: doc/source/tune/guides/overview.rst [hunk 2]
```diff
@@ -1,38 +1,177 @@
...
+~~~~~~~~~~~~~~~~
+
+- `mnist_pytorch <https://github.com/ray-project/ray/blob/master/python/ray/tune/examples/mnist_pytorch.py>`__: Converts the PyTorch MNIST example to use Tune with the function-based API. Also shows how to easily convert something relying on argparse to use Tune.
+- `mnist_pytorch_trainable <https://github.com/ray-project/ray/blob/master/python/ray/tune/examples/mnist_pytorch_trainable.py>`__: Converts the PyTorch MNIST example to use Tune with Trainable API. Also uses the HyperBandScheduler and checkpoints the model at the end.
+
+
+XGBoost Example
+~~~~~~~~~~~~~~~
+
+- `xgboost_example <https://github.com/ray-project/ray/blob/master/python/ray/tune/examples/xgboost_example.py>`__: Trains a basic XGBoost model with Tune with the function-based API and a XGBoost callback.
```
Comment (sven1977) [inline doc/source/tune/guides/overview.rst:163]:
```suggestion
- `xgboost_example <https://github.com/ray-project/ray/blob/master/python/ray/tune/examples/xgboost_example.py>`__: Trains a basic XGBoost model with Tune with the function-based API and an XGBoost callback.
```

File: doc/source/tune/guides/overview.rst [hunk 3]
```diff
@@ -1,38 +1,177 @@
...
+
+- `lightgbm_example <https://github.com/ray-project/ray/blob/master/python/ray/tune/examples/lightgbm_example.py>`__: Trains a basic LightGBM model with Tune with the function-based API and a LightGBM callback.
 
 
-.. :figure: /images/param_actor.png
+Contributed Examples
+~~~~~~~~~~~~~~~~~~~~
 
+- `pbt_tune_cifar10_with_keras <https://github.com/ray-project/ray/blob/master/python/ray/tune/examples/pbt_tune_cifar10_with_keras.py>`__: A contributed example of tuning a Keras model on CIFAR10 with the PopulationBasedTraining scheduler.
+- `genetic_example <https://github.com/ray-project/ray/blob/master/python/ray/tune/examples/genetic_example.py>`__: Optimizing the michalewicz function using the contributed GeneticSearch search algorithm with AsyncHyperBandScheduler.
```
Comment (sven1977) [inline doc/source/tune/guides/overview.rst:176]:
```suggestion
- `genetic_example <https://github.com/ray-project/ray/blob/master/python/ray/tune/examples/genetic_example.py>`__: Optimizing the michalewicz function using the contributed GeneticSearch algorithm with AsyncHyperBandScheduler.
```

**Review by sven1977 — APPROVED** [review]
LGTM. Just a few typos and reformulation suggestions.

File: doc/source/tune/guides/plot_tune-distributed.rst [hunk 4]
```diff
@@ -245,6 +272,48 @@ You should see Tune eventually continue the trials on a different worker node. S
...
+.. code-block:: python
+
+    tune.run(
+        my_trainable,
+        checkpoint_freq=10,
+        local_dir="~/path/to/results",
+        resume=True
+    )
+
+Upon a second run, this will restore the entire experiment state from ``~/path/to/results/my_experiment_name``. Importantly, any changes to the experiment specification upon resume will be ignored. For example, if the previous experiment has reached its termination, then resuming it with a new stop criterion makes no effect: the new experiment will terminate immediately after initialization. If you want to change the configuration, such as training more iterations, you can do so restore the checkpoint by setting ``restore=<path-to-checkpoint>`` - note that this only works for a single trial.
```
Comment (anabranch) [inline doc/source/tune/guides/plot_tune-distributed.rst:311]: Change middle sentence to:

> For example, if the previous experiment has reached its termination, then resuming it with a new stop criterion will not run. The new experiment will terminate immediately after initialization.

File: doc/source/tune/guides/plot_tune-distributed.rst [hunk 5]
```diff
@@ -245,6 +272,48 @@ You should see Tune eventually continue the trials on a different worker node. S
...
+
+.. warning::
+
+    This feature is still experimental, so any provided Trial Scheduler or Search Algorithm will not be preserved. Only ``FIFOScheduler`` and ``BasicVariantGenerator`` will be supported.
```
Comment (anabranch) [inline doc/source/tune/guides/plot_tune-distributed.rst:315]: what does it mean for "not be preserved" can you clarify that?

File: doc/source/tune/guides/plot_tune-tutorial.rst [hunk 6]
```diff
@@ -1,5 +1,7 @@
-Tune Walkthrough
```
Comment (anabranch) [inline doc/source/tune/guides/plot_tune-tutorial.rst:1]: file name is strange again

File: doc/source/tune/guides/plot_tune-usage.rst [hunk 7]
```diff
@@ -0,0 +1,375 @@
+.. _tune-user-guide:
+
+Tune User Guide
+===============
+
+.. warning:: Before you continue, be sure to have read :ref:`tune-60-seconds`.
+
+This document overviews features and different configurations of Tune.
```
Comment (anabranch) [inline doc/source/tune/guides/plot_tune-usage.rst:8]: This document provides an overview of the core concepts as well as some of the configurations for Tune Runs.

File: doc/source/tune/guides/plot_tune-usage.rst [hunk 8]
```diff
@@ -0,0 +1,375 @@
...
+.. code-block:: python
+
+    # Connect to an existing distributed Ray cluster
+    ray.init(address=<ray_address>)
+    tune.run(trainable, num_samples=100, resources_per_trial={"cpu": 2, "gpu": 1})
+
+Search Space (Grid/Random)
+--------------------------
+
+.. warning:: If you use a Search Algorithm, you may not be able to specify lambdas or grid search with this
```
Comment (anabranch) [inline doc/source/tune/guides/plot_tune-usage.rst:66]: this warning isn't clear to me.

Comment (anabranch) [reply doc/source/tune/guides/plot_tune-usage.rst:66]: what do you mean by lambdas?

File: doc/source/tune/guides/plot_tune-usage.rst [hunk 9]
```diff
@@ -0,0 +1,375 @@
...
+                tune.grid_search([16, 64, 256]),
+                tune.grid_search([16, 64, 256]),
+            ],
+        },
+        num_samples=10
+    )
+
+Read about this in the :ref:`Grid/Random Search API <tune-grid-random>` page.
+
+Metrics (Auto-Filled Results)
```
Comment (anabranch) [inline doc/source/tune/guides/plot_tune-usage.rst:104]: This section title is a little unclear, what is someone doing with these? Why is it important to know?

File: doc/source/tune/guides/plot_tune-usage.rst [hunk 10]
```diff
@@ -0,0 +1,375 @@
...
+            """Returns whether to stop trials and prevent new ones from starting."""
+            return self.should_stop
+
+    stopper = CustomStopper()
+    tune.run(my_trainable, stop=stopper)
+
+
+Note that in the above example the currently running trials will not stop immediately but will do so once their current iterations are complete. See the :ref:`tune-stop-ref` documentation.
+
+Logging/Tensorboard
```
Comment (anabranch) [inline doc/source/tune/guides/plot_tune-usage.rst:253]: How is this different from the logging page?

File: doc/source/tune/guides/plot_tune-usage.rst [hunk 11]
```diff
@@ -0,0 +1,375 @@
...
+
+
+Uploading Results
+-----------------
+
+If an upload directory is provided, Tune will automatically sync results to the given directory, natively supporting standard S3/gsutil commands.
+
+You can customize this to specify arbitrary storages with the ``sync_to_cloud`` argument. This argument is similar to ``sync_to_cloud`` in that it supports strings with the same replacement fields and arbitrary functions.
+
+.. code-block:: python
```
Comment (anabranch) [inline doc/source/tune/guides/plot_tune-usage.rst:341]: what would this function look like?  what does the function take in?

**Review by bllchmbrs — DISMISSED** [review]
Overall this is great, I'm sure there are a lot of annoying comments in here so ping me if there's something you want to chat about.

File: doc/source/tune/api_docs/logging.rst [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+=====================
+
+Tune has default loggers for Tensorboard, CSV, and JSON formats.
+
+Logging Path
+------------
+
+Tune will log the results of each trial to a subfolder under a specified local dir, which defaults to ``~/ray_results``.
+
+.. code-block:: bash
```
[the remark this answers is not in the document]
Comment (richardliaw) [reply doc/source/tune/api_docs/logging.rst:13]:
```suggestion
.. code-block:: python
```

File: doc/source/tune/api_docs/logging.rst [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+
+    # This logs to 2 different trial folders:
+    # ~/ray_results/trainable_name/trial_name_1 and ~/ray_results/trainable_name/trial_name_2
+    # trainable_name and trial_name are autogenerated.
+    tune.run(trainable, num_samples=2)
+
+You can specify the ``local_dir`` and ``trainable_name``:
+
+.. code-block:: bash
```
[the remark this answers is not in the document]
Comment (richardliaw) [reply doc/source/tune/api_docs/logging.rst:22]:
```suggestion
.. code-block:: python
```

File: doc/source/tune/guides/plot_tune-60-seconds.rst [hunk 14]
```diff
@@ -0,0 +1,168 @@
+.. _tune-60-seconds:
```
[the remark this answers is not in the document]
Comment (richardliaw) [reply doc/source/tune/guides/plot_tune-60-seconds.rst:1]: For https://github.com/sphinx-gallery/sphinx-gallery

File: doc/source/tune/guides/plot_tune-distributed.rst [hunk 15]
```diff
@@ -149,6 +149,33 @@ Analyze your results on TensorBoard by starting TensorBoard on the remote head m
 
 Note that you can customize the directory of results by running: ``tune.run(local_dir=..)``. You can then point TensorBoard to that directory to visualize results. You can also use `awless <https://github.com/wallix/awless>`_ for easy cluster management on AWS.
```
[the remark this answers is not in the document]
Comment (richardliaw) [reply doc/source/tune/guides/plot_tune-distributed.rst:150]: https://github.com/sphinx-gallery/sphinx-gallery

File: doc/source/tune/guides/plot_tune-60-seconds.rst (hunk 14 above)
[the remark this answers is not in the document]
Comment (richardliaw) [reply doc/source/tune/guides/plot_tune-60-seconds.rst:1]: Actually, looks like for Tune we don't need to set it as 'plot'. I'll change this..

File: doc/source/tune/guides/plot_tune-usage.rst (hunk 7 above)
Comment (richardliaw) [reply doc/source/tune/guides/plot_tune-usage.rst:8]:
```suggestion
This document provides an overview of the core concepts as well as some of the configurations for running Tune.
```

File: doc/source/rllib-training.rst [hunk 16]
```diff
@@ -461,7 +461,7 @@ Advanced Python APIs
 Custom Training Workflows
 ~~~~~~~~~~~~~~~~~~~~~~~~~
 
-In the `basic training example <https://github.com/ray-project/ray/blob/master/rllib/examples/custom_env.py>`__, Tune will call ``train()`` on your trainer once per training iteration and report the new training results. Sometimes, it is desirable to have full control over training, but still run inside Tune. Tune supports `custom trainable functions <tune-usage.html#trainable-api>`__ that can be used to implement `custom training workflows (example) <https://github.com/ray-project/ray/blob/master/rllib/examples/custom_train_fn.py>`__.
+In the `basic training example <https://github.com/ray-project/ray/blob/master/rllib/examples/custom_env.py>`__, Tune will call ``train()`` on your trainer once per training iteration and report the new training results. Sometimes, it is desirable to have full control over training, but still run inside Tune. Tune supports :ref:`custom trainable functions <trainable-docs>` that can be used to implement `custom training workflows (example) <https://github.com/ray-project/ray/blob/master/rllib/examples/custom_train_fn.py>`__.
```
[the remark this answers is not in the document]
Comment (richardliaw) [reply doc/source/rllib-training.rst:464]: For this PR, as discussed offline, I'm not going to rename across documentation

File: doc/source/tune-schedulers.rst [hunk 17]
```diff
@@ -38,7 +38,7 @@ Tune includes a distributed implementation of `Population Based Training (PBT) <
             })
```
[the remark this answers is not in the document]
Comment (richardliaw) [reply doc/source/tune-schedulers.rst:38]: No - though in the next PR, we're removing these files.

File: doc/source/tune-searchalg.rst [hunk 18]
```diff
@@ -26,14 +26,15 @@ Currently, Tune offers the following search algorithms (and library integrations
 Variant Generation (Grid Search/Random Search)
```
[the remark this answers is not in the document]
Comment (richardliaw) [reply doc/source/tune-searchalg.rst:26]: No - though in the next PR, we're removing these files

File: doc/source/tune-searchalg.rst [hunk 19]
```diff
@@ -26,14 +26,15 @@ Currently, Tune offers the following search algorithms (and library integrations
...
 ----------------------------------------------
 
-By default, Tune uses the `default search space and variant generation process <tune-usage.html#tune-search-space-default>`__ to create and queue trials. This supports random search and grid search as specified by the ``config`` parameter of ``tune.run``.
+By default, Tune uses a BasicVariantGenerator to sample trials. This supports random search and grid search as specified by the ``config`` parameter of ``tune.run``.
```
[the remark this answers is not in the document]
Comment (richardliaw) [reply doc/source/tune-searchalg.rst:29]: As discussed offline, we will change this in a separate PR.

File: doc/source/tune/api_docs/logging.rst [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+Tune automatically integrates with Viskit via the ``CSVLogger`` outputs. To use VisKit (you may have to install some dependencies), run:
+
+.. code-block:: bash
+
+    $ git clone https://github.com/rll/rllab.git
+    $ python rllab/rllab/viskit/frontend.py ~/ray_results/my_experiment
+
+The nonrelevant metrics (like timing stats) can be disabled on the left to show only the relevant ones (like accuracy, loss, etc.).
+
+.. image:: /ray-tune-viskit.png
```
[the remark this answers is not in the document]
Comment (richardliaw) [reply doc/source/tune/api_docs/logging.rst:85]: ? not sure...

[the remark this answers is not in the document]
Comment (richardliaw) [reply doc/source/tune/api_docs/logging.rst:85]: it builds as is

File: doc/source/tune/guides/plot_tune-60-seconds.rst [hunk 21]
```diff
@@ -0,0 +1,168 @@
...
+
+A gentle 60 second tour of Tune
```
[the remark this answers is not in the document]
Comment (richardliaw) [reply doc/source/tune/guides/plot_tune-60-seconds.rst:3]: hoho

File: doc/source/tune/guides/plot_tune-distributed.rst [hunk 22]
```diff
@@ -149,6 +149,33 @@ Analyze your results on TensorBoard by starting TensorBoard on the remote head m
...
+        sync_to_driver=custom_sync_func,
+    )
+
+When syncing results back to the driver, the source would be a path similar to ``ubuntu@192.0.0.1:/home/ubuntu/ray_results/trial1``, and the target would be a local path.
+This custom sync command would be also be used in node failures, where the source argument would be the path to the trial directory and the target would be a remote path. The `sync_to_driver` would be invoked to push a checkpoint to new node for a queued trial to resume.
+
+
 .. _tune-distributed-spot:
 
 Pre-emptible Instances (Cloud)
```
[the remark this answers is not in the document]
Comment (richardliaw) [reply doc/source/tune/guides/plot_tune-distributed.rst:181]: The way I see it: fault tolerance is an internal/architecture concept while "Pre-emptible Instances" is a user-defined configuration.

File: doc/source/tune/guides/plot_tune-usage.rst (hunk 10 above)
Comment (richardliaw) [reply doc/source/tune/guides/plot_tune-usage.rst:253]: I wanted to have a section that "previews"/points to the logging page. The logging page is "thick reference material", while this 'usage guide' is more of an overview/tldr.

File: doc/source/index.rst [hunk 23]
```diff
@@ -250,11 +250,9 @@ Getting Involved
    :caption: Tune
 
    tune.rst
-   Tune Guides and Tutorials <tune/generated_guides/overview.rst>
-   tune-usage.rst
+   Tutorials, Guides, Examples <tune/generated_guides/overview.rst>
```
[the remark this answers is not in the document]
Comment (richardliaw) [reply doc/source/index.rst:253]: Hm, looks like I can't do that for the toctree.... let me know if you have insights here

File: doc/source/tune/guides/plot_tune-usage.rst (hunk 11 above)
Comment (richardliaw) [reply doc/source/tune/guides/plot_tune-usage.rst:362]: GREAT CATCH i'll have to fix this

**Review by bllchmbrs — APPROVED** [review]
LGTM
